File: fincept-terminal-desktop/src/services/cache/cacheService.ts
// Unified Cache Service - Production-ready TypeScript wrapper for Rust cache backend
// Uses separate SQLite database: fincept_cache.db
// Features: Initialization checks, versioning, size limits, LRU eviction

import { invoke } from '@tauri-apps/api/core';

// ============================================================================
// Configuration
// ============================================================================

const CACHE_VERSION = '1.0.0'; // Increment when cache format changes
const CACHE_VERSION_KEY = '__cache_version__';
const MAX_CACHE_SIZE_MB = 100; // Maximum cache size in MB
const MAX_CACHE_ENTRIES = 10000; // Maximum number of cache entries
const CLEANUP_THRESHOLD = 0.9; // Start cleanup when cache is 90% full

// ============================================================================
// Types
// ============================================================================

export interface CacheEntry {
  cache_key: string;
  category: string;
  data: string;
  ttl_seconds: number;
  created_at: number;
  expires_at: number;
  last_accessed_at: number;
  hit_count: number;
  is_expired: boolean;
}

export interface TabSession {
  tab_id: string;
  tab_name: string;
  state: string;
  scroll_position: string | null;
  active_filters: string | null;
  selected_items: string | null;
  updated_at: number;
  created_at: number;
}

export interface CacheStats {
  total_entries: number;
  total_size_bytes: number;
  expired_entries: number;
  categories: CategoryStats[];
}

export interface CategoryStats {
  category: string;
  entry_count: number;
  total_size: number;
}

// TTL presets in seconds
export const TTL = {
  '1m': 60,
  '5m': 300,
  '10m': 600,
  '15m': 900,
  '30m': 1800,
  '1h': 3600,
  '6h': 21600,
  '12h': 43200,
  '24h': 86400,
  '7d': 604800,
  '30d': 2592000,
} as const;

export type TTLPreset = keyof typeof TTL;

// Backward compatibility alias
export const CacheTTL = {
  SHORT: TTL['1m'],
  MEDIUM: TTL['5m'],
  LONG: TTL['10m'],
  HOUR: TTL['1h'],
  DAY: TTL['24h'],
  WEEK: TTL['7d'],
} as const;

// Cache categories with default TTLs
export const CACHE_CATEGORIES = {
  'market-quotes': TTL['5m'],
  'market-historical': TTL['1h'],
  'news': TTL['10m'],
  'forum': TTL['5m'],
  'economic': TTL['1h'],
  'reference': TTL['24h'],
  'user-data': TTL['30d'],
  'api-response': TTL['15m'],
} as const;

export type CacheCategory = keyof typeof CACHE_CATEGORIES;

// ============================================================================
// Helper Functions
// ============================================================================

function parseTTL(ttl: number | TTLPreset): number {
  if (typeof ttl === 'number') return ttl;
  return TTL[ttl] || 600; // Default 10 minutes
}

// ============================================================================
// Cache Service Class
// ============================================================================

export class CacheService {
  private initialized = false;
  private initializing = false;
  private initPromise: Promise<boolean> | null = null;
  private cleanupInterval: ReturnType<typeof setInterval> | null = null;

  // ============================================================================
  // Initialization
  // ============================================================================

  /**
   * Initialize the cache service
   * Checks version and clears cache if version mismatch
   */
  async initialize(): Promise<boolean> {
    if (this.initialized) return true;

    // Prevent concurrent initialization
    if (this.initializing && this.initPromise) {
      return this.initPromise;
    }

    this.initializing = true;
    this.initPromise = this._doInitialize();

    try {
      const result = await this.initPromise;
      this.initialized = result;
      return result;
    } finally {
      this.initializing = false;
    }
  }

  private async _doInitialize(): Promise<boolean> {
    try {
      // Check if Rust backend is available
      const isHealthy = await this.healthCheck();
      if (!isHealthy) {
        console.error('[CacheService] Rust backend not available');
        return false;
      }

      // Check cache version
      const storedVersion = await this.getRawValue(CACHE_VERSION_KEY);
      if (storedVersion !== CACHE_VERSION) {
        console.log(`[CacheService] Cache version mismatch (${storedVersion} -> ${CACHE_VERSION}), clearing cache`);
        await this.clearAll();
        await this.setRawValue(CACHE_VERSION_KEY, CACHE_VERSION, 'system', TTL['30d']);
      }

      // Start background cleanup
      this.startBackgroundCleanup();

      console.log('[CacheService] Initialized successfully');
      return true;
    } catch (error) {
      console.error('[CacheService] Initialization failed:', error);
      return false;
    }
  }

  /**
   * Check if Rust cache backend is healthy
   */
  async healthCheck(): Promise<boolean> {
    try {
      // Try a simple operation to verify backend is working
      await invoke('cache_stats');
      return true;
    } catch (error) {
      console.error('[CacheService] Health check failed:', error);
      return false;
    }
  }

  /**
   * Ensure initialized before operations
   */
  private async ensureInitialized(): Promise<void> {
    if (!this.initialized) {
      await this.initialize();
    }
  }

  // ============================================================================
  // Core Cache Methods
  // ============================================================================

  /**
   * Get a cached value (returns null if expired or not found)
   */
  async get<T>(key: string): Promise<{ data: T; isStale: boolean } | null> {
    await this.ensureInitialized();

    try {
      const entry = await invoke<CacheEntry | null>('cache_get', { key });
      if (!entry) return null;
      return {
        data: JSON.parse(entry.data) as T,
        isStale: entry.is_expired,
      };
    } catch (error) {
      console.error('[CacheService] get error:', error);
      return null;
    }
  }

  /**
   * Get cached value even if stale (for stale-while-revalidate)
   */
  async getWithStale<T>(key: string): Promise<{ data: T; isStale: boolean } | null> {
    await this.ensureInitialized();

    try {
      const entry = await invoke<CacheEntry | null>('cache_get_with_stale', { key });
      if (!entry) return null;
      return {
        data: JSON.parse(entry.data) as T,
        isStale: entry.is_expired,
      };
    } catch (error) {
      console.error('[CacheService] getWithStale error:', error);
      return null;
    }
  }

  /**
   * Set a cached value with automatic size management
   */
  async set<T>(
    key: string,
    data: T,
    category: CacheCategory | string,
    ttl?: number | TTLPreset
  ): Promise<boolean> {
    await this.ensureInitialized();

    try {
      // Check cache size before adding
      await this.ensureCacheSize();

      const ttlSeconds = ttl
        ? parseTTL(ttl)
        : CACHE_CATEGORIES[category as CacheCategory] || 600;

      await invoke('cache_set', {
        key,
        data: JSON.stringify(data),
        category,
        ttlSeconds,
      });
      return true;
    } catch (error) {
      console.error('[CacheService] set error:', error);
      return false;
    }
  }

  /**
   * Delete a cached value
   */
  async delete(key: string): Promise<boolean> {
    await this.ensureInitialized();

    try {
      return await invoke<boolean>('cache_delete', { key });
    } catch (error) {
      console.error('[CacheService] delete error:', error);
      return false;
    }
  }

  /**
   * Get multiple cached values
   */
  async getMany<T>(keys: string[]): Promise<Map<string, T>> {
    await this.ensureInitialized();

    try {
      const entries = await invoke<CacheEntry[]>('cache_get_many', { keys });
      const result = new Map<string, T>();
      for (const entry of entries) {
        try {
          result.set(entry.cache_key, JSON.parse(entry.data) as T);
        } catch {
          // Skip invalid entries
        }
      }
      return result;
    } catch (error) {
      console.error('[CacheService] getMany error:', error);
      return new Map();
    }
  }

  // ============================================================================
  // Size Management & LRU Eviction
  // ============================================================================

  /**
   * Ensure cache size is within limits, evict LRU entries if needed
   */
  private async ensureCacheSize(): Promise<void> {
    try {
      const stats = await this.getStats();
      if (!stats) return;

      const sizeBytes = stats.total_size_bytes;
      const maxSizeBytes = MAX_CACHE_SIZE_MB * 1024 * 1024;
      const entryCount = stats.total_entries;

      // Check if we need to evict
      const sizeExceeded = sizeBytes > maxSizeBytes * CLEANUP_THRESHOLD;
      const countExceeded = entryCount > MAX_CACHE_ENTRIES * CLEANUP_THRESHOLD;

      if (sizeExceeded || countExceeded) {
        console.log(`[CacheService] Cache cleanup triggered - Size: ${(sizeBytes / 1024 / 1024).toFixed(2)}MB, Entries: ${entryCount}`);

        // First, remove expired entries
        const expiredRemoved = await this.cleanup();
        console.log(`[CacheService] Removed ${expiredRemoved} expired entries`);

        // If still over limit, use LRU eviction
        const newStats = await this.getStats();
        if (newStats && (newStats.total_size_bytes > maxSizeBytes * 0.8 || newStats.total_entries > MAX_CACHE_ENTRIES * 0.8)) {
          const targetRemoval = Math.max(
            Math.floor(newStats.total_entries * 0.2), // Remove 20% of entries
            100 // Minimum 100 entries
          );
          const lruRemoved = await this.evictLRU(targetRemoval);
          console.log(`[CacheService] LRU evicted ${lruRemoved} entries`);
        }
      }
    } catch (error) {
      console.error('[CacheService] ensureCacheSize error:', error);
    }
  }

  /**
   * Evict least recently used entries
   */
  private async evictLRU(count: number): Promise<number> {
    try {
      return await invoke<number>('cache_evict_lru', { count });
    } catch (error) {
      // Fallback: try to use pattern invalidation if LRU not available
      console.warn('[CacheService] LRU eviction not available, using cleanup fallback');
      return await this.cleanup();
    }
  }

  // ============================================================================
  // Category & Pattern Operations
  // ============================================================================

  /**
   * Invalidate all entries in a category
   */
  async invalidateCategory(category: string): Promise<number> {
    await this.ensureInitialized();

    try {
      return await invoke<number>('cache_invalidate_category', { category });
    } catch (error) {
      console.error('[CacheService] invalidateCategory error:', error);
      return 0;
    }
  }

  /**
   * Invalidate entries matching a pattern (use % for wildcard)
   */
  async invalidatePattern(pattern: string): Promise<number> {
    await this.ensureInitialized();

    try {
      return await invoke<number>('cache_invalidate_pattern', { pattern });
    } catch (error) {
      console.error('[CacheService] invalidatePattern error:', error);
      return 0;
    }
  }

  // ============================================================================
  // Cleanup & Maintenance
  // ============================================================================

  /**
   * Remove all expired entries
   */
  async cleanup(): Promise<number> {
    try {
      return await invoke<number>('cache_cleanup');
    } catch (error) {
      console.error('[CacheService] cleanup error:', error);
      return 0;
    }
  }

  /**
   * Get cache statistics
   */
  async getStats(): Promise<CacheStats | null> {
    try {
      return await invoke<CacheStats>('cache_stats');
    } catch (error) {
      console.error('[CacheService] getStats error:', error);
      return null;
    }
  }

  /**
   * Alias for getStats (backward compatibility)
   */
  async stats(): Promise<CacheStats | null> {
    return this.getStats();
  }

  /**
   * Clear entire cache (except version key)
   */
  async clearAll(): Promise<number> {
    try {
      const result = await invoke<number>('cache_clear_all');
      // Restore version key
      await this.setRawValue(CACHE_VERSION_KEY, CACHE_VERSION, 'system', TTL['30d']);
      return result;
    } catch (error) {
      console.error('[CacheService] clearAll error:', error);
      return 0;
    }
  }

  /**
   * Start background cleanup (runs every 15 minutes by default)
   */
  startBackgroundCleanup(intervalMs: number = 15 * 60 * 1000): () => void {
    // Clear any existing interval
    if (this.cleanupInterval) {
      clearInterval(this.cleanupInterval);
    }

    this.cleanupInterval = setInterval(async () => {
      try {
        const deleted = await this.cleanup();
        if (deleted > 0) {
          console.log(`[CacheService] Background cleanup: removed ${deleted} expired entries`);
        }

        // Also check size limits
        await this.ensureCacheSize();
      } catch (error) {
        console.error('[CacheService] Background cleanup error:', error);
      }
    }, intervalMs);

    return () => {
      if (this.cleanupInterval) {
        clearInterval(this.cleanupInterval);
        this.cleanupInterval = null;
      }
    };
  }

  /**
   * Stop background cleanup
   */
  stopBackgroundCleanup(): void {
    if (this.cleanupInterval) {
      clearInterval(this.cleanupInterval);
      this.cleanupInterval = null;
    }
  }

  // ============================================================================
  // Tab Session Methods
  // ============================================================================

  /**
   * Get tab session state
   */
  async getTabSession<T>(tabId: string): Promise<T | null> {
    await this.ensureInitialized();

    try {
      const session = await invoke<TabSession | null>('tab_session_get', { tabId });
      if (!session) return null;
      return JSON.parse(session.state) as T;
    } catch (error) {
      console.error('[CacheService] getTabSession error:', error);
      return null;
    }
  }

  /**
   * Save tab session state
   */
  async setTabSession<T>(
    tabId: string,
    tabName: string,
    state: T,
    scrollPosition?: string,
    activeFilters?: string,
    selectedItems?: string
  ): Promise<boolean> {
    await this.ensureInitialized();

    try {
      await invoke('tab_session_set', {
        tabId,
        tabName,
        state: JSON.stringify(state),
        scrollPosition: scrollPosition || null,
        activeFilters: activeFilters || null,
        selectedItems: selectedItems || null,
      });
      return true;
    } catch (error) {
      console.error('[CacheService] setTabSession error:', error);
      return false;
    }
  }

  /**
   * Delete tab session
   */
  async deleteTabSession(tabId: string): Promise<boolean> {
    await this.ensureInitialized();

    try {
      return await invoke<boolean>('tab_session_delete', { tabId });
    } catch (error) {
      console.error('[CacheService] deleteTabSession error:', error);
      return false;
    }
  }

  /**
   * Get all tab sessions
   */
  async getAllTabSessions(): Promise<TabSession[]> {
    await this.ensureInitialized();

    try {
      return await invoke<TabSession[]>('tab_session_get_all');
    } catch (error) {
      console.error('[CacheService] getAllTabSessions error:', error);
      return [];
    }
  }

  /**
   * Cleanup old tab sessions
   */
  async cleanupTabSessions(maxAgeDays: number = 30): Promise<number> {
    await this.ensureInitialized();

    try {
      return await invoke<number>('tab_session_cleanup', { maxAgeDays });
    } catch (error) {
      console.error('[CacheService] cleanupTabSessions error:', error);
      return 0;
    }
  }

  // ============================================================================
  // Internal Helpers
  // ============================================================================

  /**
   * Get raw string value (for internal use like version checking)
   */
  private async getRawValue(key: string): Promise<string | null> {
    try {
      const entry = await invoke<CacheEntry | null>('cache_get_with_stale', { key });
      if (!entry) return null;
      return JSON.parse(entry.data) as string;
    } catch {
      return null;
    }
  }

  /**
   * Set raw string value (for internal use like version checking)
   */
  private async setRawValue(key: string, value: string, category: string, ttlSeconds: number): Promise<boolean> {
    try {
      await invoke('cache_set', {
        key,
        data: JSON.stringify(value),
        category,
        ttlSeconds,
      });
      return true;
    } catch {
      return false;
    }
  }

  // ============================================================================
  // Backward Compatibility Methods
  // ============================================================================

  /**
   * Get raw string data from cache
   */
  async getData(key: string): Promise<string | null> {
    const result = await this.get<string>(key);
    return result?.data ?? null;
  }

  /**
   * Get cached data as JSON (backward compatibility)
   */
  async getJSON<T>(key: string): Promise<T | null> {
    const result = await this.get<T>(key);
    return result?.data ?? null;
  }

  /**
   * Set cached data as JSON (backward compatibility)
   */
  async setJSON<T>(
    key: string,
    value: T,
    category: CacheCategory | string,
    ttlSeconds: number = 300
  ): Promise<boolean> {
    return this.set(key, value, category, ttlSeconds);
  }

  // ============================================================================
  // Static Utilities
  // ============================================================================

  /**
   * Generate cache key with consistent format
   */
  static key(category: string, ...parts: (string | number)[]): string {
    return `${category}:${parts.join(':')}`;
  }

  /**
   * Generate cache key with hash for dynamic data
   */
  static keyWithHash(category: string, identifier: string, items: string[]): string {
    if (items.length === 0) return `${category}:${identifier}:empty`;
    const hash = `${items.length}:${items[0]}:${items[items.length - 1]}`;
    return `${category}:${identifier}:${hash}`;
  }

  /**
   * Get current cache version
   */
  static getVersion(): string {
    return CACHE_VERSION;
  }
}

// Export singleton instance
export const cacheService = new CacheService();
export default cacheService;

File: fincept-cpp/src/storage/cache_service.cpp
#include "cache_service.h"
#include "database.h"
#include "core/logger.h"
#include <chrono>

namespace fincept {

// ============================================================================
// CacheCategories — default TTL lookup
// ============================================================================
int64_t CacheCategories::default_ttl_for(const std::string& category) {
    static const std::map<std::string, int64_t> defaults = {
        {"market-quotes",     CacheTTL::FIVE_MIN},
        {"market-historical", CacheTTL::ONE_HOUR},
        {"news",              CacheTTL::TEN_MIN},
        {"forum",             CacheTTL::FIVE_MIN},
        {"economic",          CacheTTL::ONE_HOUR},
        {"reference",         CacheTTL::ONE_DAY},
        {"user-data",         CacheTTL::THIRTY_DAYS},
        {"api-response",      CacheTTL::FIFTEEN_MIN},
        {"system",            CacheTTL::THIRTY_DAYS},
    };
    auto it = defaults.find(category);
    return (it != defaults.end()) ? it->second : CacheTTL::TEN_MIN;
}

// ============================================================================
// CacheService — Singleton
// ============================================================================
CacheService& CacheService::instance() {
    static CacheService svc;
    return svc;
}

CacheService::~CacheService() {
    shutdown();
}

// ============================================================================
// Lifecycle
// ============================================================================
bool CacheService::initialize() {
    if (initialized_.load()) return true;

    try {
        // Ensure cache database is initialized
        auto& cache_db = db::CacheDatabase::instance();
        if (!cache_db.initialize()) {
            LOG_ERROR("CacheService", "Failed to initialize cache database");
            return false;
        }

        // Check cache version — clear if mismatched
        check_version();

        // Start background cleanup
        start_background_cleanup();

        initialized_.store(true);
        LOG_INFO("CacheService", "Initialized (version %s)", CACHE_VERSION);
        return true;
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "Initialization failed: %s", e.what());
        return false;
    }
}

void CacheService::shutdown() {
    stop_background_cleanup();
    initialized_.store(false);
}

void CacheService::check_version() {
    try {
        auto entry = db::ops::cache_get_with_stale(VERSION_KEY);
        if (!entry || entry->data != std::string("\"") + CACHE_VERSION + "\"") {
            LOG_INFO("CacheService", "Cache version mismatch, clearing cache");
            db::ops::cache_clear_all();
            // Store new version
            std::string version_data = std::string("\"") + CACHE_VERSION + "\"";
            db::ops::cache_set(VERSION_KEY, version_data, "system", CacheTTL::THIRTY_DAYS);
        }
    } catch (const std::exception& e) {
        LOG_WARN("CacheService", "Version check failed: %s", e.what());
    }
}

// ============================================================================
// Core Cache Operations
// ============================================================================
std::optional<std::string> CacheService::get(const std::string& key) {
    if (!initialized_.load()) return std::nullopt;

    try {
        auto entry = db::ops::cache_get(key);
        if (!entry) return std::nullopt;
        return entry->data;
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "get error for '%s': %s", key.c_str(), e.what());
        return std::nullopt;
    }
}

std::optional<CacheService::StaleResult> CacheService::get_with_stale(const std::string& key) {
    if (!initialized_.load()) return std::nullopt;

    try {
        auto entry = db::ops::cache_get_with_stale(key);
        if (!entry) return std::nullopt;
        return StaleResult{entry->data, entry->is_expired};
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "get_with_stale error for '%s': %s", key.c_str(), e.what());
        return std::nullopt;
    }
}

bool CacheService::set(const std::string& key, const std::string& data,
                       const std::string& category, int64_t ttl) {
    if (!initialized_.load()) return false;

    try {
        // Check size limits before inserting
        ensure_cache_size();

        int64_t ttl_seconds = (ttl > 0) ? ttl : CacheCategories::default_ttl_for(category);
        db::ops::cache_set(key, data, category, ttl_seconds);
        return true;
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "set error for '%s': %s", key.c_str(), e.what());
        return false;
    }
}

bool CacheService::remove(const std::string& key) {
    if (!initialized_.load()) return false;

    try {
        return db::ops::cache_delete(key);
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "remove error for '%s': %s", key.c_str(), e.what());
        return false;
    }
}

std::map<std::string, std::string> CacheService::get_many(const std::vector<std::string>& keys) {
    std::map<std::string, std::string> result;
    if (!initialized_.load() || keys.empty()) return result;

    try {
        auto entries = db::ops::cache_get_many(keys);
        for (auto& entry : entries) {
            result[entry.cache_key] = std::move(entry.data);
        }
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "get_many error: %s", e.what());
    }
    return result;
}

// ============================================================================
// Category & Pattern Operations
// ============================================================================
int64_t CacheService::invalidate_category(const std::string& category) {
    if (!initialized_.load()) return 0;

    try {
        return db::ops::cache_invalidate_category(category);
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "invalidate_category error: %s", e.what());
        return 0;
    }
}

int64_t CacheService::invalidate_pattern(const std::string& pattern) {
    if (!initialized_.load()) return 0;

    try {
        return db::ops::cache_invalidate_pattern(pattern);
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "invalidate_pattern error: %s", e.what());
        return 0;
    }
}

// ============================================================================
// Cleanup & Maintenance
// ============================================================================
int64_t CacheService::cleanup() {
    try {
        return db::ops::cache_cleanup();
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "cleanup error: %s", e.what());
        return 0;
    }
}

int64_t CacheService::evict_lru(int64_t count) {
    try {
        return db::ops::cache_evict_lru(count);
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "evict_lru error: %s", e.what());
        return 0;
    }
}

void CacheService::clear_all() {
    try {
        db::ops::cache_clear_all();
        // Restore version key
        std::string version_data = std::string("\"") + CACHE_VERSION + "\"";
        db::ops::cache_set(VERSION_KEY, version_data, "system", CacheTTL::THIRTY_DAYS);
        LOG_INFO("CacheService", "Cache cleared");
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "clear_all error: %s", e.what());
    }
}

CacheService::Stats CacheService::get_stats() {
    Stats result{};
    try {
        auto db_stats = db::ops::cache_stats();
        result.total_entries = db_stats.total_entries;
        result.total_size_bytes = db_stats.total_size_bytes;
        result.expired_entries = db_stats.expired_entries;
        for (auto& cat : db_stats.categories) {
            result.categories.push_back({cat.category, cat.entry_count, cat.total_size});
        }
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "get_stats error: %s", e.what());
    }
    return result;
}

// ============================================================================
// Size Management & LRU Eviction
// ============================================================================
void CacheService::ensure_cache_size() {
    try {
        auto stats = db::ops::cache_stats();

        int64_t max_size_bytes = MAX_CACHE_SIZE_MB * 1024 * 1024;
        bool size_exceeded = stats.total_size_bytes > (int64_t)(max_size_bytes * CLEANUP_THRESHOLD);
        bool count_exceeded = stats.total_entries > (int64_t)(MAX_CACHE_ENTRIES * CLEANUP_THRESHOLD);

        if (!size_exceeded && !count_exceeded) return;

        LOG_INFO("CacheService", "Cache cleanup triggered — Size: %.2f MB, Entries: %lld",
                 (double)stats.total_size_bytes / (1024.0 * 1024.0),
                 (long long)stats.total_entries);

        // First: remove expired entries
        int64_t expired_removed = db::ops::cache_cleanup();
        if (expired_removed > 0) {
            LOG_INFO("CacheService", "Removed %lld expired entries", (long long)expired_removed);
        }

        // Re-check after cleanup
        auto new_stats = db::ops::cache_stats();
        bool still_over_size = new_stats.total_size_bytes > (int64_t)(max_size_bytes * 0.8);
        bool still_over_count = new_stats.total_entries > (int64_t)(MAX_CACHE_ENTRIES * 0.8);

        if (still_over_size || still_over_count) {
            // LRU eviction — remove 20% of entries, minimum 100
            int64_t target = std::max((int64_t)(new_stats.total_entries * 0.2), (int64_t)100);
            int64_t evicted = db::ops::cache_evict_lru(target);
            LOG_INFO("CacheService", "LRU evicted %lld entries", (long long)evicted);
        }
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "ensure_cache_size error: %s", e.what());
    }
}

// ============================================================================
// Tab Sessions
// ============================================================================
std::optional<std::string> CacheService::get_tab_session(const std::string& tab_id) {
    if (!initialized_.load()) return std::nullopt;

    try {
        auto session = db::ops::tab_session_get(tab_id);
        if (!session) return std::nullopt;
        return session->state;
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "get_tab_session error: %s", e.what());
        return std::nullopt;
    }
}

bool CacheService::set_tab_session(const std::string& tab_id, const std::string& tab_name,
                                   const std::string& state, const std::string& scroll_position,
                                   const std::string& active_filters,
                                   const std::string& selected_items) {
    if (!initialized_.load()) return false;

    try {
        db::ops::tab_session_set(tab_id, tab_name, state, scroll_position,
                                 active_filters, selected_items);
        return true;
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "set_tab_session error: %s", e.what());
        return false;
    }
}

bool CacheService::delete_tab_session(const std::string& tab_id) {
    if (!initialized_.load()) return false;

    try {
        db::ops::tab_session_delete(tab_id);
        return true;
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "delete_tab_session error: %s", e.what());
        return false;
    }
}

int64_t CacheService::cleanup_tab_sessions(int64_t max_age_days) {
    try {
        return db::ops::tab_session_cleanup(max_age_days);
    } catch (const std::exception& e) {
        LOG_ERROR("CacheService", "cleanup_tab_sessions error: %s", e.what());
        return 0;
    }
}

// ============================================================================
// Background Cleanup Thread
// ============================================================================
void CacheService::start_background_cleanup(int interval_seconds) {
    stop_background_cleanup();

    cleanup_running_.store(true);
    cleanup_thread_ = std::thread([this, interval_seconds]() {
        LOG_INFO("CacheService", "Background cleanup started (interval: %ds)", interval_seconds);

        while (cleanup_running_.load()) {
            // Sleep in small increments so we can stop quickly
            for (int i = 0; i < interval_seconds && cleanup_running_.load(); i++) {
                std::this_thread::sleep_for(std::chrono::seconds(1));
            }

            if (!cleanup_running_.load()) break;

            try {
                int64_t expired = db::ops::cache_cleanup();
                if (expired > 0) {
                    LOG_INFO("CacheService", "Background cleanup: removed %lld expired entries",
                             (long long)expired);
                }

                // Also check size limits
                ensure_cache_size();

                // Cleanup old tab sessions (older than 30 days)
                db::ops::tab_session_cleanup(30);
            } catch (const std::exception& e) {
                LOG_ERROR("CacheService", "Background cleanup error: %s", e.what());
            }
        }

        LOG_INFO("CacheService", "Background cleanup stopped");
    });
}

void CacheService::stop_background_cleanup() {
    cleanup_running_.store(false);
    if (cleanup_thread_.joinable()) {
        cleanup_thread_.join();
    }
}

// ============================================================================
// Key Generation Helpers
// ============================================================================
std::string CacheService::make_key(const std::string& category, const std::string& part1) {
    return category + ":" + part1;
}

std::string CacheService::make_key(const std::string& category, const std::string& part1,
                                   const std::string& part2) {
    return category + ":" + part1 + ":" + part2;
}

std::string CacheService::make_key(const std::string& category, const std::string& part1,
                                   const std::string& part2, const std::string& part3) {
    return category + ":" + part1 + ":" + part2 + ":" + part3;
}

} // namespace fincept
